libs/vm-modules/include/vm_modules/ml/model/model_estimator.hpp

```diff
@@ -0,0 +1,152 @@
+                             math::SizeType const &inputs, math::SizeType const &hidden_nodes);
+  ChargeAmount LayerAddDenseActivation(fetch::vm::Ptr<fetch::vm::String> const &layer,
+                                       math::SizeType const &                   inputs,
+                                       math::SizeType const &                   hidden_nodes,
+                                       fetch::vm::Ptr<fetch::vm::String> const &activation);
+
+  ChargeAmount LayerAddConv(fetch::vm::Ptr<fetch::vm::String> const &layer,
+                            math::SizeType const &                   output_channels,
+                            math::SizeType const &input_channels, math::SizeType const &kernel_size,
+                            math::SizeType const &stride_size);
+  ChargeAmount LayerAddConvActivation(fetch::vm::Ptr<fetch::vm::String> const &layer,
+                                      math::SizeType const &                   output_channels,
+                                      math::SizeType const &                   input_channels,
+                                      math::SizeType const &                   kernel_size,
+                                      math::SizeType const &                   stride_size,
+                                      fetch::vm::Ptr<fetch::vm::String> const &activation);
+
+  ChargeAmount CompileSequential(fetch::vm::Ptr<fetch::vm::String> const &loss,
+                                 fetch::vm::Ptr<fetch::vm::String> const &optimiser);
+
+  ChargeAmount CompileSimple(fetch::vm::Ptr<fetch::vm::String> const &        optimiser,
+                             fetch::vm::Ptr<vm::Array<math::SizeType>> const &in_layers);
+
+  ChargeAmount Fit(vm::Ptr<vm_modules::math::VMTensor> const &data,
+                   vm::Ptr<vm_modules::math::VMTensor> const &labels,
+                   ::fetch::math::SizeType const &            batch_size);
+
+  ChargeAmount Evaluate();
+
+  ChargeAmount Predict(vm::Ptr<vm_modules::math::VMTensor> const &data);
+
+  ChargeAmount SerializeToString();
+
+  ChargeAmount DeserializeFromString(fetch::vm::Ptr<fetch::vm::String> const &model_string);
+
+  bool SerializeTo(serializers::MsgPackSerializer &buffer);
+
+  bool DeserializeFrom(serializers::MsgPackSerializer &buffer);
+
+  SizeType GetPaddedSizesSum();
+  SizeType GetSizesSum();
+
+private:
+  struct State
+  {
+    // Model
+    ChargeAmount forward_pass_cost{0};
+    ChargeAmount backward_pass_cost{0};
+
+    // Optimiser
+    SizeType weights_size_sum{0};
+    SizeType weights_padded_size_sum{0};
+
+    ChargeAmount optimiser_step_impact{0};
+
+    SizeType last_layer_size{0};
+    SizeType ops_count{0};
+
+    // serialization
+    bool SerializeTo(serializers::MsgPackSerializer &buffer);
+    bool DeserializeFrom(serializers::MsgPackSerializer &buffer);
+  };
+
+  VMObjectType &model_;
+  State         state_;
+
+  // Forward
+  static constexpr double FORWARD_DENSE_INPUT_COEF  = 0.142857142857143;
```

Comment:
these must not be doubles - that's a potential source of indeterminacy. FixedPoints should be fine right?